nomt/src/beatree/ops/update/leaf_stage.rs

```diff
@@ -285,42 +333,85 @@ fn try_answer_left_neighbor(
         },
     };
 
-    // We send the left side information it needs to deduce the first node following the previous
-    // high point of its range. note that `self.low` and `left.high` are kept equal.
-
+    // We send the left side information it needs to deduce the nodes to use next.
+    // A modified leaf could be used as a node by the left worker, or we provide a new
+    // high range to let the left worker use new fresh leaves
     let mut found_next_node = false;
-    let mut next_separator = worker_params.range.low.clone();
+    let mut found_unchanged_range = false;
+    let mut new_high_range = worker_params.range.low.clone();
 
-    // We take the first rewritten node in the range.
-    let take = leaf_changes
+    let take = leaves_tracker
         .inner
         .iter()
-        .take_while(|(_, entry)| {
-            if found_next_node {
-                return false;
+        .take_while(|(_, entry)| match entry {
+            LeavesTrackerEntry::UnchangedRange(range) => {
+                found_unchanged_range = true;
+                new_high_range = range.high;
+                false
             }
-
-            found_next_node = entry.inserted.is_some();
-            next_separator = entry.next_separator.clone();
-            true
+            LeavesTrackerEntry::ChangedLeaf(entry) if entry.inserted.is_some() => {
+                found_next_node = true;
+                new_high_range = entry.next_separator.clone();
+                false
+            }
+            _ => true,
         })
         .count();
 
-    let consumed_whole_range = next_separator == worker_params.range.high;
-
-    if !found_next_node && !consumed_whole_range {
+    // Keep the request pending only if no inserted node has been found or any unchanged range,
+    // and if the worker has not finished with the right worker
+    if !(found_next_node || found_unchanged_range) && !has_finished_workload {
         *pending_request = Some(request);
         return;
     }
 
-    // update our low range to match.
-    worker_params.range.low = next_separator;
-
-    let changed = (0..take)
-        .map(|_| leaf_changes.inner.pop_first().unwrap())
+    // `self.low` and `left.high` are kept equal
+    worker_params.range.low = new_high_range;
+
+    let mut deleted = (0..take)
+        .map(|_| {
+            leaves_tracker
+                .inner
+                .pop_first()
+                .and_then(|(key, entry)| match entry {
+                    // SAFETY: the existance and variant of `take` leaves_tracker items was previously checked
+                    LeavesTrackerEntry::ChangedLeaf(changed_entry) => Some((key, changed_entry)),
+                    _ => None,
+                })
+                .unwrap()
```

Comment:
why does this `unwrap` still hold?

Reply:
my bad, during some refactoring I forgot to move and update a comment a little bit above, instead of `SAFETY` it should be `UNWRAP` and it explains why the `unwrap` holds